### PR TITLE
[Feature] Filling data on professional skills in a user profile

### PR DESCRIPTION
**Pull request by weesom**
Link to the ticket:
https://trello.com/c/Ay9rXEca/12-feature-filling-data-on-professional-skills-in-a-user-profile

What it does:
Allows to choose and update skill levels for User
...

DB changes:
UserSkillLevel
Skill
+ some validation fixes added to User

### CODE REVIEW

**Comment by gogochikago** [comment]
and update action like
def update 
  @skills = Skills.where(user_id: params[:user_id])
  if @skills.update
    redirect_to user_path
  else 
    render :edit

**Comment by weesom** [comment]
[[Feature] Filling data on professional skills in a user profile](https://trello.com/c/Ay9rXEca/12-feature-filling-data-on-professional-skills-in-a-user-profile)

File: config/routes.rb [hunk 1]
```diff
@@ -7,6 +7,12 @@
   post '/login', to: 'sessions#create'
   delete '/logout', to: 'sessions#destroy'
 
-  resources :users
+  # resources :users
```
Comment (gogochikago) [inline config/routes.rb]: You can delete it if you don't use.

File: db/seeds.rb [hunk 2]
```diff
@@ -1,12 +1,73 @@
...
+  'MySQL',
+  'NoSQL',
+  'ETL / OLAP / BI / Data Mining'
+]
+
+databases.each do |datab|
+  Skill.where(name: datab, sphere: 'Database').first_or_create
+end
+
+# User.create()
```
Comment (gogochikago) [inline db/seeds.rb]: You can delete it if you don't use.

File: app/models/user.rb [hunk 3]
```diff
@@ -5,17 +5,21 @@ class User < ApplicationRecord
   include Authenticatable
   attr_accessor :remember_token
 
+  has_many :user_skill_levels
```
Comment (gogochikago) [inline app/models/user.rb]: I think you need to add a dependency. If you remove the user, then that his skills are also deleted.

File: app/views/shared/header/_header.html.slim [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+          = link_to 'Profile', profile_path(current_user)
+        span class="dropdown-item"
+          = link_to 'Settings', '#'
+        span class="dropdown-item"
+          = link_to 'Log out', logout_path, method: 'delete'
+        span class="dropdown-item"
+          = link_to 'Skills', user_skill_level_path(current_user)
+  - else
+    span class="btn btn-outline-primary"
+      = link_to 'Log in', login_path
```
Comment (gogochikago) [inline app/views/shared/header/_header.html.slim]: We have navbar instead header.

File: app/views/users/new.html.slim [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+      = f.label :password
+      = f.password_field :password, class: 'form-control'
+      br
+
+      = f.label :password_confirmation, 'Confirmation'
+      = f.password_field :password_confirmation, class: 'form-control'
+      br
+
+      = f.submit 'Create account', class: 'btn btn-success'
+      br
```
Comment (gogochikago) [inline app/views/users/new.html.slim]: This file is not used

File: spec/controllers/profiles_controller_spec.rb [hunk 6]
```diff
@@ -3,159 +3,172 @@
...
+		context 'when logged out' do
+			it 'does not update the record in the database and redirect to root' do
+				patch :update, params: { user_id: profile2.id, profile: valid_attribute }
+				expect(profile.reload.telephone).to eq('375291111111')
+				expect(response).to redirect_to(root_path)
+			end
+		end
+	end
 end
-# rubocop:enable all
```
Comment (gogochikago) [inline spec/controllers/profiles_controller_spec.rb]: After merge pr with factory_bot configuration, rewrite this with factories

File: spec/controllers/user_skill_level_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+      get :edit, params: { user_id: user_skill_level.user_id }
+      expect(response).to be_successful
+    end
+
+    it 'renders user_skill_levels#edit template' do
+      get :edit, params: { user_id: user_skill_level.user_id }
+      expect(response).to render_template(:edit)
+    end
+  end
+end
```
Comment (gogochikago) [inline spec/controllers/user_skill_level_controller_spec.rb]: After merge pr with factory_bot configuration, rewrite this with factories

File: app/assets/javascripts/user_skill_level.coffee [hunk 8]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (labs-ruby) [inline app/assets/javascripts/user_skill_level.coffee]: Remove coffeescript generator at all

File: app/controllers/profiles_controller.rb [hunk 9]
```diff
@@ -3,17 +3,22 @@
 # Controller for profile
 class ProfilesController < ApplicationController
   def show
-    @profile = authorize Profile.find(params[:id])
+    @profile = Profile.find(params[:user_id])
+    authorize @profile
+    @user = current_user
+    @skills = Skill.all
```
Comment (labs-ruby) [inline app/controllers/profiles_controller.rb]: Not good idea to request all skills at once from DB.
What if there'll be much of them?

File: app/controllers/user_skill_levels_controller.rb [hunk 10]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+# Controller
```
Comment (labs-ruby) [inline app/controllers/user_skill_levels_controller.rb]: Adjust your rubocop config with `  Documentation:
    Enabled: false`

This is a useless comment

File: app/models/user.rb [hunk 11]
```diff
@@ -6,6 +6,9 @@ class User < ApplicationRecord
   attr_accessor :remember_token
   has_one :profile, dependent: :destroy
 
+  has_many :user_skill_levels, dependent: :destroy
```
Comment (labs-ruby) [inline app/models/user.rb]: I would suggest to move skill level to a User namespace,
then, adjust your association name to be just `has_many :skill_levels, dependent: :destroy`
see `user.user_skill_levels` vs `user.skill_levels`

File: app/models/user_skill_level.rb [hunk 12]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+# Model of UserSkillLevel
+class UserSkillLevel < ApplicationRecord
```
Comment (labs-ruby) [inline app/models/user_skill_level.rb]: Move to User namespace

File: spec/controllers/profiles_controller_spec.rb [hunk 13]
```diff
@@ -3,159 +3,172 @@
...
+		end
 
-      context 'with invalid params' do
-        it 'does not update the record in the database' do
-          patch :update, params: { id: profile.id, profile: invalid_attribute }
-          expect(profile.reload.telephone).to eq('375291111111')
-        end
-      end
-    end
+		context 'when the user tries to change not his profile' do
```
Comment (labs-ruby) [inline spec/controllers/profiles_controller_spec.rb]: What happened with indentation?

Comment (weesom) [reply spec/controllers/profiles_controller_spec.rb]: Converted to tabs instead of spaces.

File: app/controllers/profiles_controller.rb (hunk 9 above)
Comment (weesom) [reply app/controllers/profiles_controller.rb]: That is actually the idea to show all skills that user can have. They are generated in 'seed.rb' and admin would be able to add more (possibly).

**Pull request merged**